src/runtime/opengl/opengl_module.cc

```diff
@@ -0,0 +1,291 @@
+  stream->Write(fmt_);
+  stream->Write(fmap_);
+  stream->Write(ToJson(shaders_));
+}
+
+const gl::Program& OpenGLModuleNode::GetProgram(
+    const std::string& func_name) const {
+  auto it = programs_.find(func_name);
+  if (it == programs_.end()) {
+    LOG(FATAL) << "Cannot find program";
+  }
+  return it->second;
+}
+
+const OpenGLShader& OpenGLModuleNode::GetShader(
+    const std::string& func_name) const {
+  auto it = shaders_.find(func_name);
+  if (it == shaders_.end()) {
+    LOG(FATAL) << "Cannot find shader";
+  }
+  return it->second;
+}
+
+const FunctionInfo& OpenGLModuleNode::GetFunctionInfo(
+    const std::string& func_name) const {
+  auto it = fmap_.find(func_name);
+  if (it == fmap_.end()) {
+    LOG(FATAL) << "Cannot find shader";
+  }
+  return it->second;
+}
+
+OpenGLWrappedFunc::OpenGLWrappedFunc(
+    OpenGLModuleNode* m,
+    std::shared_ptr<ModuleNode> sptr,
+    std::string func_name,
+    std::vector<size_t> arg_size,
+    const std::vector<std::string>& thread_axis_tags)
+    : m_(m), sptr_(std::move(sptr)), func_name_(std::move(func_name)),
+      arg_size_(std::move(arg_size)) {
+  LOG(INFO) << "OpenGLWrappedFunc(" << func_name_ << ", "
+            << "nargs = " << arg_size_.size() << ", "
+            << "nthread_axis_tags = " << thread_axis_tags.size() << ")";
+  for (auto& a : arg_size_) {
+    LOG(INFO) << a;
+  }
+  for (auto& t : thread_axis_tags) {
+    LOG(INFO) << t;
+  }
+
+  thread_axis_cfg_.Init(arg_size_.size(), thread_axis_tags);
+}
+
+void OpenGLWrappedFunc::operator()(TVMArgs args, TVMRetValue* rv,
+                                   void** void_args) const {
+  LOG(INFO) << "OpenGLWrappedFunc::operator()";
+
+  auto &shader = m_->GetShader(func_name_);
+  auto &program = m_->GetProgram(func_name_);
+  auto &func_info = m_->GetFunctionInfo(func_name_);
+
+  size_t nargs = shader.arg_kinds.size();
+
+  std::vector<std::tuple<std::string, TVMType, void*>> uniforms;
+  std::vector<std::pair<std::string, gl::Texture*>> inputs;
+  gl::Texture* output = nullptr;
+  for (size_t i = 0; i != nargs; ++i) {
+    auto name = shader.arg_names.at(i);
+    auto kind = shader.arg_kinds.at(i);
+    auto type = func_info.arg_types.at(i);
+    switch (kind) {
+      case OpenGLArgKind::kUniform: {
+        uniforms.emplace_back(name, type, void_args[i]);
+        break;
+      }
+      case OpenGLArgKind::kInputTexture: {
+        CHECK_EQ(type.code, kHandle) << "Type is not handle?";
+        auto texture = *static_cast<gl::Texture**>(void_args[i]);
+        inputs.emplace_back(name, texture);
+        break;
+      }
+      case OpenGLArgKind::kOutputTexture: {
+        CHECK_EQ(type.code, kHandle) << "Type is not handle?";
+        output = *static_cast<gl::Texture**>(void_args[i]);
+        break;
+      }
+      default: {
+        LOG(FATAL) << "Invalid OpenGLArgKind";
+      }
+    }
+  }
+
+  m_->workspace().Render(program, uniforms, inputs, output);
```

Comment:
I remember gl have API that is like setArg, which allows you to direct set Uniform and input of each object. Let us consider use that one as it might be more efficient than allocating a vector for each of these

Reply:
I'm a little bit reluctant to do that. The reason is that all `gl*` API's as well as certain OpenGL concepts are encapsulated within `OpenGLWorkspace`.